File: PI_Controller.m
% BE 441/541: Instrumentation and Controls
% PID controller tuning and implementation from a step response

clear all; % Clears all defined MATLAB variables

% Define PI controller parameters
Kp = 2.0; % Change this value to desired proportional gain
Ki = 0.5; % Change this value to desired integral gain

% Define other parameters
dt = 0.5; % Sampling interval in seconds
t_end = 10 * 60; % Time to record in seconds (converted from 10 minutes)
T_Safe = 120; % Safe temperature threshold in °F
SP = 100; % Setpoint temperature in °F

% Create an Arduino object. Replace 'ComX' with the correct COM port (e.g., 'COM3')
a = arduino('ComX', 'Nano');

% Record the start time
startTime = datetime('now');

% Initialize variables for PI control
integral = 0;
previous_error = 0;

% Initialize time tracking and data logging
Delta_t = seconds(datetime('now') - startTime); % Time elapsed in seconds
Count = 0; % Initialize counter for data points

% Main loop to control temperature until `t_end` is reached
while Delta_t < t_end
    % Read the current temperature in °F from the sensor connected to 'A0'
    PV = readVoltage(a, 'A0') * 180 + 32; % Adjust pin if necessary
    
    % Calculate the error
    error = SP - PV;
    
    % Update integral term
    integral = integral + error * dt;
    
    % Calculate control output
    u = Kp * error + Ki * integral;
    
    % Limit the control output to be between 0 and 1
    u = max(0, min(1, u));
    
    % Apply the control output to the heater
    writePWMDutyCycle(a, 'D3', u);
    
    % Check if the temperature exceeds the safe threshold
    if PV > T_Safe
        writePWMDutyCycle(a, 'D3', 0); % Stop the PWM signal if temperature is unsafe
        disp('Temperature exceeded safe level...Process stopped.')
        break; % Terminate the process
    end
    
    % Update the time elapsed
    Delta_t = seconds(datetime('now') - startTime);
    
    % Increment the counter and store data
    Count = Count + 1;
    t(Count) = Delta_t; % Store the time elapsed
    T(Count) = PV; % Store the temperature reading
    
    % Display the current time and temperature
    disp([t(Count), T(Count)]);
    
    % Pause for the sampling interval
    pause(dt);
end

% Save the recorded data to a .mat file
save('temperature_data.mat', 't', 'T');

% Plot the recorded temperature data
plot(t, T);
xlabel('Time (s)');
ylabel('Temperature (°F)');
title('PI Control Response');
grid on;

File: PI_Controller.cpp
// BE 441/541: Instrumentation and Controls
// PI controller implementation 

#define pwmPin  3     // Replace with the appropriate pin
#define tempPin  A0   // Replace with the appropriate analog pin
float x = 100; // Setpoint temperature in °F
const float Kp = 2.0;  // Proportional gain
const float Ki = 0.5;  // Integral gain
const float dt = 0.5;  // Sampling interval in seconds
const int t_end = 10 * 60; // Time to record in seconds (converted from 10 minutes)
const float T_Safe = 120; // Safe temperature threshold in °F


unsigned long startTime;
unsigned long currentTime;
unsigned long elapsedTime;
float integral = 0;
float previous_error = 0;
int count = 0;

void setup() {
  Serial.begin(9600);
  pinMode(pwmPin, OUTPUT);
  Serial.println("Time(s), Temperature(°F)");
  startTime = millis(); // Record the start time
}

void loop() {
  currentTime = millis();
  elapsedTime = (currentTime - startTime) / 1000; // Time elapsed in seconds

  if (elapsedTime < t_end) {
    float voltage = analogRead(tempPin) * (5.0 / 1023.0); // Read the current voltage
    float y = voltage * 180 + 32; // Convert voltage to temperature in °F

    // Calculate the error
    float error = x - y;

    // Update integral term
    integral += error * dt;

    // Calculate control output
    float u = Kp * error + Ki * integral;

    // Limit the control output to be between 0 and 1
    u = constrain(u, 0, 1);

    // Apply the control output to the heater
    analogWrite(pwmPin, u * 255);

    // Check if the temperature exceeds the safe threshold
    if (y > T_Safe) {
      analogWrite(pwmPin, 0); // Stop the PWM signal if temperature is unsafe
      Serial.println("Temperature exceeded safe level...Process stopped.");
    }

    // Log the time and temperature
    Serial.print(elapsedTime);
    Serial.print(", ");
    Serial.println(y);

    delay(dt * 1000); // Pause for the sampling interval
  }
}
